Refactor how kernel binaries are stored and how we write them to workers for FD

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/11014

### Problem description
To enable linking within/across sub cmds for kernel binaries, need to reorganize how we iterate/write binaries from being binary/address major, core range minor to core range major, binary/address minor

### What's changed

- Packed all kernels of a program into a single buffer. This reduces wasted memory as well as removing our binary duplication across kernel groups.
- Change iteration to be core range major, binary minor. Currently reviewing HW issues with linking when we an additional transaction on a different command buffer/VC, so we don't enable linking in this PR.
- Additional change to uplift write to pcie from device to use new noc apis, and not reprogram NOC coord

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

tt-aho committed:
#0: Update process_write_host_h to new dispatch NOC api to avoid reprogramming NOC coord

aliuTT commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -828,99 +828,96 @@ void EnqueueProgramCommand::assemble_device_commands(
+            std::visit(
+                [&](auto&& cores) {
+                    using T = std::decay_t<decltype(cores)>;
+                    if constexpr (std::is_same_v<T, CoreRange>) {
```
Is this generic for WORKER and ETH? What if the EthernetKernel uses `CoreRange((0,0) to (0,2))`? I don't think that's asserted anywhere. It will cause this check to bug out for ETH, maybe add to the CreateEthernetKernel an assert that cores have to be CoreCoord for WH? In BH it's probably ok, depending on how `get_noc_multicast_encoding` gets uplifted.

tt-aho replied:
This is generic for workers and eth and is how main is currently set up. The code for this part hasn't changed.

This is not affected by how the user creates their kernel as this comes from how we encoded the coords in `tt_metal/impl/program/program.cpp`.

```
for (KernelGroup &kernel_group : this->get_kernel_groups(CoreType::WORKER)) {
         std::vector<pair<transfer_info_cores, uint32_t>> dst_noc_multicast_info = extract_dst_noc_multicast_info<std::set<CoreRange>>(
             device, kernel_group.core_ranges.ranges(), kernel_group.get_core_type());
```
The code above generates the CoreRanges consumed by the code you highlighted, whereas the code below generates the CoreCoords. command_queue.cpp is agnostic to whether it is eth or worker.
```
for (KernelGroup &kernel_group : this->get_kernel_groups(CoreType::ETH)) {
        vector<pair<transfer_info_cores, uint32_t>> dst_noc_unicast_info =
            extract_dst_noc_unicast_info(kernel_group.core_ranges.ranges(), kernel_group.get_core_type());
```

aliuTT replied:
Thanks, I forgot my own work 🐒

tt-aho committed:
#11014: Refactor program kernel bins to be packed into 1 buffer
Reorganize kernel data so that we iterate over core range major, binary minor for kernel groups

abhullar-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -828,99 +816,94 @@ void EnqueueProgramCommand::assemble_device_commands(
+                    uint32_t relayed_bytes =
+                        align(kg_transfer_info.lengths[kernel_idx], HostMemDeviceCommand::PROGRAM_PAGE_SIZE);
+                    // length_adjust needs to be aligned to NOC_DRAM_ALIGNMENT
+                    uint16_t length_adjust = uint16_t(relayed_bytes - kg_transfer_info.lengths[kernel_idx]);
```
does this need to be explicitly aligned?

tt-aho replied:
No that was a legacy comment from when we were aligning it to dram alignment for some reason. I removed the alignment a while back but missed removing the comment. I'll remove.

abhullar-tt commented on tt_metal/impl/dispatch/kernels/cq_dispatch.cpp:
```diff
@@ -195,28 +196,33 @@ void process_write_host_h() {
         completion_queue_push_back(npages);
-        noc_async_writes_flushed();
-        block_noc_writes_to_clear[rd_block_idx] += div_up(xfer_size, NOC_MAX_BURST_SIZE);
+        uint32_t num_noc_packets_written = div_up(xfer_size, NOC_MAX_BURST_SIZE) + 1;
```
the +1 is for the write to host? If so can you add a comment explaining that

tt-aho replied:
Yeah +1 is from completion_queue_push_back writing the updated ptr to host. I removed the counter increments from there to minimize add/stores. Will add comment
